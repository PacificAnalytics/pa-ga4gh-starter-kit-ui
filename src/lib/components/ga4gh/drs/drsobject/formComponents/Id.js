import React from 'react';
import { 
    Typography,
    FormControl,
    TextField,
    Grid,
    Button
} from '@material-ui/core';
import { v4 as uuidv4 } from 'uuid';

const Id = props => {
    return(
        <Grid container spacing={4} alignItems='center' justify='space-between'>
            <Grid item xs>
                <FormControl fullWidth>
                    <TextField
                        id='id'
                        label='Id'
                        margin='normal'
                        name='id'
                        type='text'
                        value={props.id}
                        InputProps={{readOnly: props.readOnly}} 
                        onChange={e => props.setId(e.target.value)}
                        helperText='Unique identifier for this DRS Object (UUID recommended), cannot be modified later.'
                    />
                </FormControl>
            </Grid>
            {/* button to generate id  */}
            {props.readOnly
                ? null
                : (
                    <Grid item xs={2} align='right'>
                        <Button variant='contained' color='primary' 
                        onClick={() =>  props.setId(uuidv4())}>
                            <Typography variant='button'>Generate ID</Typography>
                        </Button>
                    </Grid>     
                )
            }
        </Grid>
    );
}

export default Id;
