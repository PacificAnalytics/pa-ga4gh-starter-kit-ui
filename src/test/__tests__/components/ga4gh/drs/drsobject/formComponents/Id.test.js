import React from 'react';
import renderer from 'react-test-renderer';
import {render, screen} from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import userEvent from '@testing-library/user-event';
import {mockUpdateScalar} from '../../../../../../resources/MockFunctions';
import Id from '../../../../../../../lib/components/ga4gh/drs/drsobject/formComponents/Id';

afterEach(() => {
    mockUpdateScalar.mockClear();
})

test('SHOW and EDIT <Id /> should handle populated id field', () => {
    let {container} = render(<Id readOnly={true} id='1234567890' setId={mockUpdateScalar} />);
    expect(container.firstChild).toMatchSnapshot();
    expect(screen.getByLabelText('Id')).toBeInTheDocument();
    let idField = screen.getByRole('textbox');
    expect(idField).toHaveValue('1234567890');
    // id field is not editable and "Generate UUID" button is not displayed
    expect(screen.queryByRole('button')).not.toBeInTheDocument();
    expect(idField).toHaveAttribute('readOnly');
    userEvent.type(idField, 'new test value');
    expect(mockUpdateScalar.mock.calls.length).toBe(0);
});

test('NEW <Id /> should handle empty id field', () => {
    let {container} = render(<Id readOnly={false} id='' setId={mockUpdateScalar} />);
    expect(container.firstChild).toMatchSnapshot();
    expect(screen.getByLabelText('Id')).toBeInTheDocument();
    let idField = screen.getByRole('textbox');
    expect(idField).toHaveValue('');
    // id field is editable and can be populated using "Generate UUID" button
    userEvent.type(idField, 'abc123');
    expect(mockUpdateScalar).toHaveBeenCalled();
    mockUpdateScalar.mockClear();
    let generateIdButton = screen.getByRole('button');
    expect(generateIdButton).toBeInTheDocument();
    userEvent.click(generateIdButton);
    expect(mockUpdateScalar.mock.calls.length).toBe(1);
});

test('NEW <Id /> should handle populated id field', () => {
    let {container} = render(<Id readOnly={false} id='1234567890' setId={mockUpdateScalar} />);
    expect(container.firstChild).toMatchSnapshot();
    expect(screen.getByLabelText('Id')).toBeInTheDocument();
    let idField = screen.getByRole('textbox');
    expect(idField).toHaveValue('1234567890');
    // id field is editable and can be populated using "Generate UUID" button
    userEvent.type(idField, 'abc123');
    expect(mockUpdateScalar).toHaveBeenCalled();
    mockUpdateScalar.mockClear();
    let generateIdButton = screen.getByRole('button');
    expect(generateIdButton).toBeInTheDocument();
    userEvent.click(generateIdButton);
    expect(mockUpdateScalar.mock.calls.length).toBe(1);
});