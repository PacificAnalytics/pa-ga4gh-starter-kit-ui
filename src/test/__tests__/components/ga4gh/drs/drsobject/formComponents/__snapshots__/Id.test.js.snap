// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NEW <Id /> should handle empty id field 1`] = `
<div
  class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-4 MuiGrid-align-items-xs-center MuiGrid-justify-content-xs-space-between"
>
  <div
    class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true"
  >
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth"
    >
      <div
        class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
          data-shrink="false"
          for="id"
          id="id-label"
        >
          Id
        </label>
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
        >
          <input
            aria-describedby="id-helper-text"
            aria-invalid="false"
            class="MuiInputBase-input MuiInput-input"
            id="id"
            name="id"
            type="text"
            value=""
          />
        </div>
        <p
          class="MuiFormHelperText-root"
          id="id-helper-text"
        >
          Unique identifier for this DRS Object (UUID recommended), cannot be modified later.
        </p>
      </div>
    </div>
  </div>
  <div
    align="right"
    class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-2"
  >
    <button
      aria-label="generate-id-button"
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        <span
          class="MuiTypography-root MuiTypography-button"
        >
          Generate ID
        </span>
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;

exports[`NEW <Id /> should handle populated id field 1`] = `
<div
  class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-4 MuiGrid-align-items-xs-center MuiGrid-justify-content-xs-space-between"
>
  <div
    class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true"
  >
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth"
    >
      <div
        class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
          data-shrink="true"
          for="id"
          id="id-label"
        >
          Id
        </label>
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
        >
          <input
            aria-describedby="id-helper-text"
            aria-invalid="false"
            class="MuiInputBase-input MuiInput-input"
            id="id"
            name="id"
            type="text"
            value="1234567890"
          />
        </div>
        <p
          class="MuiFormHelperText-root MuiFormHelperText-filled"
          id="id-helper-text"
        >
          Unique identifier for this DRS Object (UUID recommended), cannot be modified later.
        </p>
      </div>
    </div>
  </div>
  <div
    align="right"
    class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-2"
  >
    <button
      aria-label="generate-id-button"
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        <span
          class="MuiTypography-root MuiTypography-button"
        >
          Generate ID
        </span>
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;

exports[`SHOW and EDIT <Id /> should handle populated id field 1`] = `
<div
  class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-4 MuiGrid-align-items-xs-center MuiGrid-justify-content-xs-space-between"
>
  <div
    class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true"
  >
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth"
    >
      <div
        class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
          data-shrink="true"
          for="id"
          id="id-label"
        >
          Id
        </label>
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
        >
          <input
            aria-describedby="id-helper-text"
            aria-invalid="false"
            class="MuiInputBase-input MuiInput-input"
            id="id"
            name="id"
            readonly=""
            type="text"
            value="1234567890"
          />
        </div>
        <p
          class="MuiFormHelperText-root MuiFormHelperText-filled"
          id="id-helper-text"
        >
          Unique identifier for this DRS Object (UUID recommended), cannot be modified later.
        </p>
      </div>
    </div>
  </div>
</div>
`;
